lua_tests/gambiarra/gambiarra.lua

```diff
@@ -0,0 +1,296 @@
+  elseif e == 'finish' then
+    print("######## "..e.."ed "..test.." tests")
+  else
+    print(e.." "..test)
+  end
+end
+
+--[[
+if equal returns true
+if different returns {msg = "<reason>"}
+  this will be handled spechially by ok and nok
+--]]
+local function deepeq(a, b)
+  local function notEqual(m)
+    return { msg=m }
+  end
+  
+  -- Different types: false
+  if type(a) ~= type(b) then return notEqual("type 1 is "..type(a)..", type 2 is "..type(b)) end
+  -- Functions
+  if type(a) == 'function' then
+    if string.dump(a) == string.dump(b) then
+      return true
+    else 
+      return notEqual("functions differ")
+    end
+  end
+  -- Primitives and equal pointers
+  if a == b then return true end
+  -- Only equal tables could have passed previous tests
+  if type(a) ~= 'table' then return notEqual("different "..type(a).."s expected "..a.." vs. "..b) end
+  -- Compare tables field by field
+  for k,v in pairs(a) do
+    if b[k] == nil then return notEqual("key "..k.."only contained in left part") end
+    local result = deepeq(v, b[k])
+    if type(result) == 'table' then return result end
+  end
+  for k,v in pairs(b) do
+    if a[k] == nil then return notEqual("key "..k.."only contained in right part") end
+    local result = deepeq(a[k], v)
+    if type(result) == 'table' then return result end
+  end
+  return true
+end
+
+-- Compatibility for Lua 5.1 and Lua 5.2
+local function args(...)
+  return {n=select('#', ...), ...}
+end
+
+local function spy(f)
+  local s = {}
+  setmetatable(s, {__call = function(s, ...)
+    s.called = s.called or {}
+    local a = args(...)
+    table.insert(s.called, {...})
+    if f then
+      local r
+      r = args(pcall(f, (unpack or table.unpack)(a, 1, a.n)))
+      if not r[1] then
+        s.errors = s.errors or {}
+        s.errors[#s.called] = r[2]
+      else
+        return (unpack or table.unpack)(r, 2, r.n)
+      end
+    end
+  end})
+  return s
+end
+
+local function assertok(handler, name, invert, cond, msg)
+  local errormsg
+  if type(cond) == 'table' and cond.msg then
+    errormsg = cond.msg
+    cond = false
+  end
+  if not msg then
+    -- debug.getinfo() does not exist in NodeMCU
+    -- msg = debug.getinfo(2, 'S').short_src..":"..debug.getinfo(2, 'l').currentline
+    msg = debug.traceback()
+    msg = msg:match("([^\t]*): in[^\n]*\n\t[^\n]*in function '[ROM.]*pcall'")
+    msg = msg:match(".-([^\\/]*)$") -- cut off path of filename
+  end
+
+  if invert then
+    cond = not cond
+  end
+  if cond then
+    handler('pass', name, msg)
+  else
+    handler('fail', name, msg, errormsg)
+    error('_*_TestAbort_*_')
+  end
+end
+
+local function fail(handler, name, func, expected, msg)
+  if not msg then
+    -- debug.getinfo() does not exist in NodeMCU
+    -- msg = debug.getinfo(2, 'S').short_src..":"..debug.getinfo(2, 'l').currentline
+    msg = debug.traceback()
+    msg = msg:match("([^\t]*): in[^\n]*\n\t[^\n]*in function '[ROM.]*pcall'")
```

Comment:
It seems odd to use `[ROM.]*` with a character class.  Is this a Lua-ism that the more apparent regex `(ROM\.)*` isn't a Pattern?

Reply:
I also tried your suggestion at first but it doesn't match. So I am using this slightly too relaxed regexp.

Reply:
PiL contains the surprising (to me) assertion that "Unlike some other systems, in Lua a modifier can only be applied to a character class; there is no way to group patterns under a modifier. For instance, there is no pattern that matches an optional word (unless the word has only one letter)."  So this is probably fine, but does anything go wrong if you leave off the `[ROM.]*` part, since it's being considered optional with `*`, or make it use `+` instead?

Reply:
The problem is that Lua 5.1 shows 'pcall' on the stack whereas Lua 5.3 shown 'ROM.pcall' in the same place.
So I tried to create a pattern that matches both.
In the meantime I remembered that Terry wrote to have the full debug implementation in Lua 5.3, so I changed the implementation back to using debug.getinfo and can remove this part of the pattern again.
Will checkin these days. I am working on freeing the result array as soon as possible instead of building up a second array holding the actual outputs and compare them to the expected at the end of the test only. Output in case of error will change for the selftest but who cares.

Reply:
Thanks for explaining!